### ISSUE TITLE
Substrate Considerations

### ISSUE THREAD

**Issue by kayabaNerve** [serai-dex/serai#379]
The following is a collection of loose notes/comments/complaints/TODOs regarding Substrate, specifically the one present in https://github.com/serai-dex/polkadot-sdk.

- Usage of a Base58-based scheme. We should use bech32m.
- Presence of type metadata (`primitives/metadata-ir`).
  
  We don't want to remove metadata entirely. We do want to remove it from the on-chain WASM blob. That probably means we remove the getters which return it from the blob, and that'll cause it to be linked out.
  
  We can do size comparisons on pre-remove-getters and post-remove-getters to verify it's no longer being included.
- Schnorrkel context of "substrate", instead of "serai".
- Events are stored in a single value (in order to achieve efficient removal), creating a very-high cost to read as any read must be of all events (or at least the first `n`).
- GRANDPA uses Ed25519, not sr25519.
- I believe lack of compile time checks for storage key conflicts (yet will have to check).
- parity-db is pulled in when it may not be production viable yet (#386).
- Lack of context used by sign/verify functions.
- Remove the wasm blob from the state.
- deposit_event panics on having 4 billion events. It should return an error.
- We may want to replace the state root with solely an events root.

This supersedes #28, which has gotten out of hand. This issue will be edited in place, instead of appended to.

**Comment by akildemir** [comment]
Always ignored the `__Ignore` in the events or calls, because it says "ignore"  🤣

**Comment by kayabaNerve** [comment]
Yeah, that's the way to handle it if it exists lol. I'm just curious if we can remove it entirely/find the originally documented reason for its existence (presumably to make the enum non-exhaustive, yet there's a dedicated attribute for that (https://github.com/rust-lang/rust-clippy/issues/10550 provides great context)).

**Comment by kayabaNerve** [comment]
We don't want to remove metadata entirely. We do want to remove it from the on-chain WASM blob. That probably means we remove the getters which return it from the blob, and that'll cause it to be linked out.

We can do size comparisons on pre-remove-getters and post-remove-getters to verify it's no longer being included.

**Comment by kayabaNerve** [comment]
https://github.com/serai-dex/serai/commit/776e417fd2313bda1a8522aa14a3c4d5c92553d3 owned the ABI and effectively solves every topic here.
